ODIN_II/SRC/odin_error.cpp

```diff
@@ -107,8 +106,29 @@ void _log_message(odin_error error_type, long column, long line_number, long fil
             fprintf(stderr, "\n");
     }
 
-    print_culprit_line(column, line_number, file);
+    if (file >= 0 && (size_t)file < include_file_names.size()
+        && line_number >= 0) {
+        print_culprit_line(column, line_number, include_file_names[file].first.c_str());
+    }
 
     fflush(stderr);
-    _verbose_assert(!fatal_error, "", function_file_name, function_line, function_name);
+    _verbose_assert(!fatal_error, NULL, function_file_name, function_line, function_name);
+}
+
+void _verbose_assert(bool condition, const char* condition_str, const char* odin_file_name, long odin_line_number, const char* odin_function_name) {
+    fflush(stdout);
+    if (!condition) {
+        fprintf(stderr, "%s:%ld: %s: ", odin_file_name, odin_line_number, odin_function_name);
+        if (condition_str) {
+            // We are an assertion, print the condition that failed and which line it occurred on
+            fprintf(stderr, "Assertion %s failed\n", condition_str);
+            // -1 since __LINE__ starts from 1
+            print_culprit_line(-1, odin_line_number - 1, odin_file_name);
```

Comment:
Would this be more valuable if it printed a couple of lines of context as well?

Reply:
yes it would! would you mind adding that?

Reply:
`/home/dast/vtr-verilog-to-routing/ODIN_II/SRC/odin_util.cpp:106: open_file: Assertion false failed
104: FILE* open_file(const char* file_name, const char* open_type) {
105:     FILE* opened_file = fopen(file_name, open_type);
106:     oassert(false);
         ^~~~
107:     if (opened_file == NULL) {
108:         error_message(UTIL, -1, -1, "cannot open file: %s\n", file_name);`

Reply:
Looks awesome! Thanks for that!